src/opensignals/data/provider.py

```diff
@@ -0,0 +1,257 @@
+from abc import ABC, abstractmethod
+from concurrent import futures
+from datetime import datetime, date, time
+import logging
+import shutil
+
+from dateutil.relativedelta import relativedelta, FR
+import pandas as pd
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+AWS_BASE_URL = 'https://numerai-signals-public-data.s3-us-west-2.amazonaws.com'
+SIGNALS_UNIVERSE = f'{AWS_BASE_URL}/latest_universe.csv'
+SIGNALS_TICKER_MAP = f'{AWS_BASE_URL}/signals_ticker_map_w_bbg.csv'
+SIGNALS_TARGETS = f'{AWS_BASE_URL}/signals_train_val_bbg.csv'
+
+
+class Provider(ABC):
+    """Common base class for (daily) stock price data"""
+
+    @staticmethod
+    def get_tickers():
+        ticker_map = pd.read_csv(SIGNALS_TICKER_MAP)
+        ticker_map = ticker_map.dropna(subset=['yahoo'])
+        logger.info(f'Number of eligible tickers: {ticker_map.shape[0]}')
+
+        if ticker_map['yahoo'].duplicated().any():
+            num = ticker_map["yahoo"].duplicated().values().sum()
+            raise Exception(
+                f'Found duplicated {num}'
+                ' yahoo tickers'
+            )
+
+        if ticker_map['bloomberg_ticker'].duplicated().any():
+            num = ticker_map["bloomberg_ticker"].duplicated().values().sum()
+            raise Exception(
+                f'Found duplicated {num}'
+                ' bloomberg_ticker tickers'
+            )
+
+        return ticker_map
+
+    @staticmethod
+    def get_ticker_data(db_dir):
+        ticker_data = pd.DataFrame({
+            'bloomberg_ticker': pd.Series([], dtype='str'),
+            'date': pd.Series([], dtype='datetime64[ns]')
+        })
+        if len(list(db_dir.rglob('*.parquet'))) > 0:
+            ticker_data = pd.read_parquet(db_dir)
+
+        num = ticker_data.bloomberg_ticker.unique().shape[0]
+        logger.info(f'Retrieving data for {num} tickers from the database')
+
+        return ticker_data
+
+    @staticmethod
+    def get_ticker_missing(ticker_data,
+                           ticker_map,
+                           last_friday=datetime.today() - relativedelta(weekday=FR(-1))):
```

Comment:
The statement defining the default value for `last_friday` is executed upon definition time, not at call time. That could eventually cause some weird behavior. The saver way is defaulting to None then add some `if last_friday is None: last_friday = ... `

Reply:
That's a good one. I think I did a faithful refactor, but this indeed is one of Python's amazing gotchas.